Docs restructuring and minor edits (#129)
spatialos/online-services: docs/content/services-packages/overview.md

# Services & packages overview
<%(TOC)%>

The [Online Services repository](http://github.com/spatialos/online-services) contains:

* services & packages - see below.
* configuration examples - see [`SampleClient`](https://github.com/spatialos/online-services/tree/master/services/csharp/SampleClient) and [`SampleMatcher`](https://github.com/spatialos/online-services/tree/master/services/csharp/SampleMatcher).

> Each Online Service is an executable which runs in the cloud and each Online Service package is a discrete set of functionality which offers you set up as part of a Online Service.

Author: Oh, whoops - I spotted the typo here previously, but forgot to update it 🤦‍♀ 
I think I just need to get rid of "offers", so that it becomes "which you set up".

> Each Online Service is an executable which runs in the cloud and each Online Service package is a discrete set of functionality which offers you set up as part of a Online Service.

Author: I imagine the writer initially planned to say "which offers you X, Y, and Z" and then changed their mind :)

## Services

Deployable cloud services for matchmaking and authentication.

### Gateway (including matchmaking)

> For matchmaking, you can use the Gateway service. This consists of:

Reviewer: When reading this, I thought "Why not call it the Matchmaking service, then?". This reads slightly weirdly to me.

Author: I know what you mean. However, I know there are reasons behind the naming (it's not _just_ matchmaking) and I don't think we're able to change it at this point.

* **Gateway**

>     The client-facing interface to the matchmaking system. Exposes two gRPC APIs: the Gateway API and a [long-running operations](https://github.com/googleapis/googleapis/blob/master/google/longrunning/operations.proto) API.

Author: @petemounce Could you cast an eye over this section (`#Gateway (including matchmaking)`) and check that I've used "service" and "API" correctly? Unfortunately GitHub thinks this is a whole new page, so the specific changes don't show up. You can view a preview here: https://docs-staging.improbable.io/metagame/0.1.0/index

Reviewer 2: I think this makes sense as a way to avoid the "service" confusion.

Author: Great, thanks!

    - [Gateway proto definition](http://github.com/spatialos/online-services/tree/master/services/proto/gateway/gateway.proto)

>     - [Long-running operations proto definition](http://github.com/spatialos/online-services/tree/master/services/proto/google/longrunning/operations.proto)

Reviewer 2: We're scrapping it in the next release anyway - will fix docs when that PR lands

    - [Service implementation (C#)](http://github.com/spatialos/online-services/tree/master/services/csharp/Gateway)<br><br>

* **Gateway-internal**

>     Used by the Gateway, Gateway-internal is the matcher-facing interface to the matchmaking service. Exposes a `GatewayInternal` gRPC API - with the default configuration this is only exposed to other services on the Kubernetes cluster.

Author: I don't know what the second sentence is trying to say, as the grammar is a little dodgy. Any ideas?

>     Used by the Gateway, Gateway-internal is the matcher-facing interface to the matchmaking service. Exposes a `GatewayInternal` gRPC API - with the default configuration this is only exposed to other services on the Kubernetes cluster.

Reviewer 2: You're right, this is odd.

```suggestion
    Gateway-internal is the Matcher-facing interface to the matchmaking service. It exposes a `GatewayInternal` gRPC API - with the default configuration, this is only exposed to other services on the Kubernetes cluster.
```

Reviewer 2: Is this any better?

Author: Yep, thanks!

Is "Matcher" capitalised because it's the name of A Thing? Like "Gateway"? I can see mentions of "a matcher" and "matchers" in the docs, but that's all. I don't mind either way, as long as (a) we only use caps for names, and (b) we're consistent.

Reviewer 2: It's the name of a thing! I'll leave it up to you whether or not we should capitalise it :)

Author: Given how we talk about it here:
![image](https://user-images.githubusercontent.com/32169118/67498201-fba55100-f676-11e9-90c3-4c9a8876df79.png)

I feel like we shouldn't capitalise, since it looks like there's no single "Matcher". But LMK if I'm misunderstanding. (We can also leave it for now and follow up later if it comes up again.)

Reviewer 2: It's a concept like a worker IMO. We should follow the same capitalisation strategy as a worker :)

    - [Proto definition](http://github.com/spatialos/online-services/tree/master/services/proto/gateway/gateway_internal.proto)
    - [Service implementation (C#)](http://github.com/spatialos/online-services/tree/master/services/csharp/GatewayInternal)<br><br>

* **Party & invite**

>     Also used by the Gateway, this is a separate, but related, service to the matchmaking system. Provides operations for the management of parties and invites to those parties. Exposes `Party` and `Invite` gRPC APIs.

Author: Is "companion" a common/standard term that people would recognise? Thinking it might prove tricky when it comes to translation. (I certainly didn't want to google it!)

Also, I'm not sure why you've changed the second sentence from active to passive - is there any reason in particular?

>     Also used by the Gateway, this is a separate, but related, service to the matchmaking system. Provides operations for the management of parties and invites to those parties. Exposes `Party` and `Invite` gRPC APIs.

Reviewer: 2nd sentence - personal style I guess. I thought the first form was ambiguous; could have meant the gateway or the matchmaking system.

_I_ think companion is a decent word here. Subjective.

>     Also used by the Gateway, this is a separate, but related, service to the matchmaking system. Provides operations for the management of parties and invites to those parties. Exposes `Party` and `Invite` gRPC APIs.

Reviewer 2: I'm fine with companion

>     Also used by the Gateway, this is a separate, but related, service to the matchmaking system. Provides operations for the management of parties and invites to those parties. Exposes `Party` and `Invite` gRPC APIs.

Author: I can't see why changing "Provides operations..." to "Operations...are provided" would make the subject of the sentence clearer - I think both are equally ambiguous/clear - so I'm going to keep "Provides...".

Regarding "companion", I think it's probably fine for native English speakers, and fine for non-native speakers **_if_** it's a standard tech term. If it _**isn't**_, I think it's safer to stick with "separate, but related", and avoid using the term "companion" in an unusual sense.

## Packages

Discrete sets of functionality which you can set up as part of a Online Service.

All packages are namespaced with `Improbable.OnlineServices.*`. You can find these on NuGet, but they're also included in this repository and imported as `ProjectReference`s in the example services.

### Base.Server

A generic C# gRPC server. Provides convenience methods for mounting services and adding interceptors, as well as as logging and support for exporting metrics to a [Prometheus](https://prometheus.io/) instance.

This package doesn't include anything Improbable-specific; you can use it for any C# server.

- [Source & documentation](http://github.com/spatialos/online-services/tree/master/services/csharp/Base.Server/)
- [`Base.Server` package on NuGet](https://www.nuget.org/packages/Improbable.OnlineServices.Base.Server)

### Base.Matcher

A base class for implementing a Gateway [Matcher]({{urlRoot}}/content/services-packages/gateway.md#matchers).

- [Source](http://github.com/spatialos/online-services/tree/master/services/csharp/Base.Matcher/)
- [`Base.Matcher` package on NuGet](https://www.nuget.org/packages/Improbable.OnlineServices.Base.Matcher)

### Common

> A collection of classes and utilities for building online services. This includes our data model, database client libraries, Platform SDK, PIT interceptors and more. Include this library if you're building an online service for a SpatialOS game.

Reviewer: PIT interceptor? Elaborate the PIT acronym?

Author: @JoelOtter assuming PIT means Player Identity Token here? If so, I'll elaborate.
